File: commonsense_guide_to_data_strucures_and_algorithms/edition2/ch03/p3.js
// Use Big O Notation to describe the time complexity of the following function:

function chessboardSpace(numberOfGrains) {
  let chessboardSpaces = 1;
  let placedGrains = 1;

  while (placedGrains < numberOfGrains) {
    placedGrains *= 2;
    chessboardSpaces += 1;
  }
  
  return chessboardSpaces;
}

// O(logN)

File: commonsense_guide_to_data_strucures_and_algorithms/edition2/ch04/p3.py
# Use Big O Notation to describe the time complexity of the following function. It finds the greatest product of any pair of two numbers within a given array:

def greatestProduct(array):
  greatestProductSoFar = array[0] * array[1]

  for i, iVal in enumerate(array):
    for j, jVal in enumerate(array):
      if i != j and iVal * jVal > greatestProductSoFar:
        greatestProductSoFar = iVal * jVal

  return greatestProductSoFar

  # O(N^2)